### PR TITLE
Fix updating empty set of credentials

### PR DESCRIPTION
**Pull request by sophokles73**
This addresses #2735 

The registry implementations failed to reject a request to update a
device's empty set of credentials, e.g. right after the device has been
created, if the request contained a secret having an ID. In fact, the
registry implementations would accept such a request even if the secrets
did not contain a password/key etc at all but only the identifier.

The registries have been changed to now explicitly check if a device's
credentials on record are empty before merging in updated credentials.

### CODE REVIEW

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CredentialsDto.java [hunk 1]
```diff
@@ -201,25 +242,30 @@ public void createMissingSecretIds() {
...
         Objects.requireNonNull(otherCredentialsDto);
 
-        Optional.ofNullable(otherCredentialsDto.getCredentials())
-                .ifPresent(credentialsToMerge -> this.getData()
-                        .forEach(credential -> findCredentialByIdAndType(credential.getAuthId(), credential.getType(),
-                                credentialsToMerge)
-                                        .ifPresent(credential::merge)));
+        final var credentialsToMerge = otherCredentialsDto.getCredentials();
+        if (credentialsToMerge == null || credentialsToMerge.isEmpty()) {
+            throw new IllegalArgumentException("DTO does not contain any credentials");
```
Comment (kaniyan) [inline services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CredentialsDto.java:258]: With respect to issue #2735, FMPOV the issue is that we are able to update with secret-ids that do not exist already provided the existing credentials list is empty. If the existing credentials list is not empty, then currently we send the error message `other credential has no secret with id: xyz`. But with this fix, in the case of empty list of existing credentials the error message is `DTO does not contain any credentials` and IMHO it is bit misleading. Shouldn't we keep it consistent with the other one?

**Review by kaniyan — CHANGES_REQUESTED** [review]

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CredentialsDto.java (hunk 1 above)
Comment (sophokles73) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CredentialsDto.java:258]: What is misleading about it? We could still change it to e.g.
> no credentials on record for device

Or did you have anything specific in mind?

Comment (kaniyan) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CredentialsDto.java:258]: FMPOV the issue is that a user is trying to update with secret-ids that do not exist. In this scenario, IMHO an error message such as  `no secrets are available for the given secret ids` or `other credential has no secret with id: xyz` (in consistent with the other scenario). Also `no credentials on record for device` suits too.

Comment (sophokles73) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CredentialsDto.java]: I have rebased and updated the error message ...

**Review by kaniyan — APPROVED** [review]
LGTM

**Pull request merged**